### PR TITLE
Implement SEQ Instruction

### PR DESCRIPTION
**Pull request by nikkischnelle**
This is the first instruction I'm implementing, so I'm just gonna open this as a draft for discussion.

### CODE REVIEW

**Comment by flamion** [comment]
Looks good, but 2 things:
- I would like more descriptive naming of the tests instead of just "XXXMatch" and "XXXNoMatch"
- SEQ is just a rename of the CMP - the instructions both do the same thing. The story related to this pull request even has it as a requirement that there are no separate implementations for both

**Comment by nikkischnelle** [comment]
I didn't see that Compare was implemented, because it isn't named CMP.kt
So Compare is already implemented, but why does the Parser call Seq for both the CMP and SEQ token?
image

And why do both Compare.kt and SEQ.kt even exist if the parser just points to _one_ implementation?

**Comment by flamion** [comment]
> I didn't see that Compare was implemented, because it isn't named CMP.kt So Compare is already implemented, but why does the Parser call Seq for both the CMP and SEQ token? image
> 
> And why do both Compare.kt and SEQ.kt even exist if the parser just points to _one_ implementation?

Because after #163 is done, CMP should just be an alias for SEQ. We can continue with this implementation, however then CMP should be just an alias for SEQ. SEQ is preferred as it fits in with the other jumping-instructions and came later as a "replacement"

**Comment by nikkischnelle** [comment]
But because both files exists, there was an existing implementation of CMP which we could have used instead, so we just did double the work. It wasn't much work, but should still be avoided in the future.
And in light of your second comment, I don't really understand the second point of your first comment

> - SEQ is just a rename of the CMP - the instructions both do the same thing. The story related to this pull request even has it as a requirement that there are no separate implementations for both

because that requirement is fulfilled by this implementation (because it had already been fullfilled beforehand)

**Comment by flamion** [comment]
> But because both files exists, there was an existing implementation of CMP which we could have used instead, so we just did double the work. It wasn't much work, but should still be avoided in the future. And in light of your second comment, I don't really understand the second point of your first comment
> 
> > * SEQ is just a rename of the CMP - the instructions both do the same thing. The story related to this pull request even has it as a requirement that there are no separate implementations for both
> 
> because that requirement is fulfilled by this implementation (because it had already been fullfilled beforehand)

We had decided while planning that an implementation for CMP should still exist, even if it just proxies an SEQ (for example by just extending SEQ but not overriding anything) - We could also just move the CMP implementation to the SEQ implementation, maybe we can discuss this in the interpreter team

**Comment by nikkischnelle** [comment]
That should have been made clear in the issue. It says

> - [ ] SEQ instruction has been implemented according to spec
> - [ ] Made sure that CMP is just an alias for SEQ and doesn't exist separately

which, to me, very clearly reads that CMP is an alias and SEQ contains the actual implementation, not the other way around.

**Comment by nikkischnelle** [comment]
Ok after a small side-discussion with @flamion :
- use this implementation
- make Cmp.kt a :socks: for the Seq.kt implementation
- adjust Parser so it calls Cmp.kt for `CMP` Tokens and Seq.kt for `SEQ` Tokens
- adjust the tests accordingly

did I miss something?

**Comment by flamion** [comment]
> Ok after a small side-discussion with @flamion :
> 
>     * use this implementation
> 
>     * nuke current Compare.kt
> 
>     * future issues will contain more detail about the context they were created in
> 
> 
> did I miss something?

Nothing missed from what I can tell

File: Containerfile.dev
Comment (flamion) [inline Containerfile.dev:1]: Why is this in this PR? This doesn't have anything to do with the SEQ Instruction?

**Review by flamion — CHANGES_REQUESTED** [review]

File: Containerfile.dev
Comment (nikkischnelle) [reply Containerfile.dev:1]: why is it even existent is the question. I drafted this yesterday but tought I scrapped it... was in a different branch entirely :D

**Review by flamion — APPROVED** [review]

**Pull request merged**